In [1]:
#Tamim Mahmud From Daffodil International University
#19/06/2025

Softmax কেন লাগে?

ধরো তুমি একটা multi-class classification করতেছো — যেমন 3টা class:

🐶 Dog


🐱 Cat


🐰 Rabbit


তোমার model এর শেষ লেয়ারে যা বের হয়, তাকে বলে logits বা raw scores। যেমন:


scores = [2.0, 1.0, 0.1]
এইগুলো কোনো probability না — বরং "confidence" এর মত।

Softmax এই raw scores গুলোকে probability তে convert করে দেয় — এমনভাবে যেন:

সবগুলা value থাকে 0 থেকে 1 এর মধ্যে

সবগুলা যোগ করলে হয় 1

In [2]:
import torch
import torch.nn.functional as F

# Model's raw scores (logits) for 3 classes
scores = torch.tensor([2.0, 1.0, 0.1])

# Apply softmax
probabilities = F.softmax(scores, dim=0)

print("Probabilities:", probabilities)
print("Sum:", probabilities.sum())



Probabilities: tensor([0.6590, 0.2424, 0.0986])
Sum: tensor(1.0000)


In [5]:
import torch
import torch.nn.functional as F
score=torch.tensor([1.0,3.0,2.0,4.0,6.0])

probabilities=F.softmax(score,dim=0) #Converts scores → probabilities

print("Probabilities:",probabilities)
print("Sum:",probabilities.sum())

Probabilities: tensor([0.0056, 0.0411, 0.0151, 0.1118, 0.8263])
Sum: tensor(1.0000)


Step 1 : proti ta score er e power x beboahr kori

setp 2: shob gulo ke jug kori

step3: প্রতিটা স্কোর ভাগ করি ঐ যোগফল দিয়ে

step 4: ja ashe shb jug krle asn ashbe 1

In [6]:
  #CrossEntropyLoss কি?
#PyTorch এ CrossEntropyLoss মানে:

#🔥 Softmax + Log + Negative Log Likelihood (NLL) — সব এক লাইনে করে ফেলে!

scores = [2.0, 1.0, 0.1]

Softmax output = [0.659, 0.242, 0.099]

now ,
: Take Negative Log of True class

Loss=−log(0.659)=0.417

এটাই CrossEntropyLoss!

In [8]:
#softmax->Probability বের করে
#-log(true_class_prob)->Loss হিসাব করে
#CrossEntropyLoss->	সবকিছু একসাথে করে দেয়

In [11]:
import torch
import torch.nn as nn
import torch.nn.functional as F

# raw model output
scores = torch.tensor([2.0, 10.0, 0.1])  # logits
labels = torch.tensor([0])             # correct class index

# Method 1: manual softmax + log + loss
softmax = F.softmax(scores, dim=0)
log_softmax = torch.log(softmax)
loss_manual = -log_softmax[labels]
print("Manual Loss:", loss_manual.item())

# Method 2: use CrossEntropyLoss directly (preferred)
loss_fn = nn.CrossEntropyLoss()
# CrossEntropyLoss expects input as [batch_size, num_classes]
# so we need to reshape scores
scores_batch = scores.unsqueeze(0)  # shape: [1, 3]
loss = loss_fn(scores_batch, labels)
print("CrossEntropyLoss:", loss.item())


Manual Loss: 8.000385284423828
CrossEntropyLoss: 8.000385284423828


In [15]:
import torch
import torch.nn as nn
import torch.nn.functional as F

s=torch.tensor([20.0,4.0,6.0,8.0])

softmax=F.softmax(s,dim=0)
print("Probabilities:",softmax)
print("Sum:",softmax.sum())


labels=torch.tensor([0])
logg=torch.log(softmax)
loss_manual=-logg[labels]

print("Loss Manual:",loss_manual.item())

loss_fn=nn.CrossEntropyLoss()
score_batch=s.unsqueeze(0)
loss=loss_fn(score_batch,labels)
print("CrossEntropyLoss:",loss.item())

Probabilities: tensor([9.9999e-01, 1.1253e-07, 8.3152e-07, 6.1442e-06])
Sum: tensor(1.0000)
Loss Manual: 7.152582838898525e-06
CrossEntropyLoss: 7.152531907195225e-06


PyTorch-এ nn.CrossEntropyLoss batch of samples নিয়ে কাজ করতে পারে।
এটি প্রতিটা স্যাম্পলের জন্য loss বের করে, তারপর সেগুলোর average নেয়

In [23]:
import torch
import torch.nn as nn
import torch.nn.functional as F

# Logits: raw model output (batch_size = 3, num_classes = 3)
logits = torch.tensor([[2.0, 1.0, 0.1],   # True class: 0
                       [1.0, 3.0, 0.1],   # True class: 1
                       [0.2, 0.3, 0.5]])  # True class: 2

labels = torch.tensor([0, 1, 2])

# PyTorch built-in CrossEntropyLoss
loss_fn = nn.CrossEntropyLoss()
loss = loss_fn(logits, labels)
print("🔹 CrossEntropy Loss:", loss.item())

# Manually calculating softmax per row (dim=1)
softmax = F.softmax(logits, dim=1)
print("🔸 Probabilities:\n", softmax)

# Get probability of correct class for each sample
probs_of_correct = softmax[torch.arange(3), labels]
print("✅ Prob of correct class per sample:", probs_of_correct)

# Manual loss = -log(prob)
loss_manual = -torch.log(probs_of_correct)
print("📉 Manual Loss per sample:", loss_manual)

# Mean of all sample losses (same as nn.CrossEntropyLoss)
manual_loss_mean = loss_manual.mean()
print("📊 Final Manual Loss:", manual_loss_mean.item())


🔹 CrossEntropy Loss: 0.5103718638420105
🔸 Probabilities:
 tensor([[0.6590, 0.2424, 0.0986],
        [0.1137, 0.8401, 0.0462],
        [0.2894, 0.3199, 0.3907]])
✅ Prob of correct class per sample: tensor([0.6590, 0.8401, 0.3907])
📉 Manual Loss per sample: tensor([0.4170, 0.1743, 0.9398])
📊 Final Manual Loss: 0.5103718638420105


In [29]:
import torch
import torch.nn as nn
import numpy as np
def softmax(x):
  return np.exp(x) / np.sum(np.exp(x),axis=0)

x=np.array([2.0,1.0,0.1])

re=softmax(x)
print("softmax:",re)
print("sum:",re.sum())

x = torch.tensor([2.0, 1.0, 0.1])
outputs = torch.softmax(x, dim=0)
print('softmax torch:', outputs)


softmax: [0.65900114 0.24243297 0.09856589]
sum: 1.0
softmax torch: tensor([0.6590, 0.2424, 0.0986])


In [1]:
import torch
import torch.nn as nn
import numpy as np

#
#        -> 2.0              -> 0.65
# Linear -> 1.0  -> Softmax  -> 0.25   -> CrossEntropy(y, y_hat)
#        -> 0.1              -> 0.1
#
#     scores(logits)      probabilities
#                           sum = 1.0
#

# Softmax applies the exponential function to each element, and normalizes
# by dividing by the sum of all these exponentials
# -> squashes the output to be between 0 and 1 = probability
# sum of all probabilities is 1
def softmax(x):
    return np.exp(x) / np.sum(np.exp(x), axis=0)

x = np.array([2.0, 1.0, 0.1])
outputs = softmax(x)
print('softmax numpy:', outputs)

x = torch.tensor([2.0, 1.0, 0.1])
outputs = torch.softmax(x, dim=0) # along values along first axis
print('softmax torch:', outputs)

# Cross entropy
# Cross-entropy loss, or log loss, measures the performance of a classification model
# whose output is a probability value between 0 and 1.
# -> loss increases as the predicted probability diverges from the actual label
def cross_entropy(actual, predicted):
    EPS = 1e-15
    predicted = np.clip(predicted, EPS, 1 - EPS)
    loss = -np.sum(actual * np.log(predicted))
    return loss # / float(predicted.shape[0])

# y must be one hot encoded
# if class 0: [1 0 0]
# if class 1: [0 1 0]
# if class 2: [0 0 1]
Y = np.array([1, 0, 0])
Y_pred_good = np.array([0.7, 0.2, 0.1])
Y_pred_bad = np.array([0.1, 0.3, 0.6])
l1 = cross_entropy(Y, Y_pred_good)
l2 = cross_entropy(Y, Y_pred_bad)
print(f'Loss1 numpy: {l1:.4f}')
print(f'Loss2 numpy: {l2:.4f}')

# CrossEntropyLoss in PyTorch (applies Softmax)
# nn.LogSoftmax + nn.NLLLoss
# NLLLoss = negative log likelihood loss
loss = nn.CrossEntropyLoss()
# loss(input, target)

# target is of size nSamples = 1
# each element has class label: 0, 1, or 2
# Y (=target) contains class labels, not one-hot
Y = torch.tensor([0])

# input is of size nSamples x nClasses = 1 x 3
# y_pred (=input) must be raw, unnormalizes scores (logits) for each class, not softmax
Y_pred_good = torch.tensor([[2.0, 1.0, 0.1]])
Y_pred_bad = torch.tensor([[0.5, 2.0, 0.3]])
l1 = loss(Y_pred_good, Y)
l2 = loss(Y_pred_bad, Y)

print(f'PyTorch Loss1: {l1.item():.4f}')
print(f'PyTorch Loss2: {l2.item():.4f}')

# get predictions
_, predictions1 = torch.max(Y_pred_good, 1)
_, predictions2 = torch.max(Y_pred_bad, 1)
print(f'Actual class: {Y.item()}, Y_pred1: {predictions1.item()}, Y_pred2: {predictions2.item()}')

# allows batch loss for multiple samples

# target is of size nBatch = 3
# each element has class label: 0, 1, or 2
Y = torch.tensor([2, 0, 1])

# input is of size nBatch x nClasses = 3 x 3
# Y_pred are logits (not softmax)
Y_pred_good = torch.tensor(
    [[0.1, 0.2, 3.9], # predict class 2
    [1.2, 0.1, 0.3], # predict class 0
    [0.3, 2.2, 0.2]]) # predict class 1

Y_pred_bad = torch.tensor(
    [[0.9, 0.2, 0.1],
    [0.1, 0.3, 1.5],
    [1.2, 0.2, 0.5]])

l1 = loss(Y_pred_good, Y)
l2 = loss(Y_pred_bad, Y)
print(f'Batch Loss1:  {l1.item():.4f}')
print(f'Batch Loss2: {l2.item():.4f}')

# get predictions
_, predictions1 = torch.max(Y_pred_good, 1)
_, predictions2 = torch.max(Y_pred_bad, 1)
print(f'Actual class: {Y}, Y_pred1: {predictions1}, Y_pred2: {predictions2}')

# Binary classification
class NeuralNet1(nn.Module):
    def __init__(self, input_size, hidden_size):
        super(NeuralNet1, self).__init__()
        self.linear1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.linear2 = nn.Linear(hidden_size, 1)

    def forward(self, x):
        out = self.linear1(x)
        out = self.relu(out)
        out = self.linear2(out)
        # sigmoid at the end
        y_pred = torch.sigmoid(out)
        return y_pred

model = NeuralNet1(input_size=28*28, hidden_size=5)
criterion = nn.BCELoss()

# Multiclass problem
class NeuralNet2(nn.Module):
    def __init__(self, input_size, hidden_size, num_classes):
        super(NeuralNet2, self).__init__()
        self.linear1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.linear2 = nn.Linear(hidden_size, num_classes)

    def forward(self, x):
        out = self.linear1(x)
        out = self.relu(out)
        out = self.linear2(out)
        # no softmax at the end
        return out

model = NeuralNet2(input_size=28*28, hidden_size=5, num_classes=3)
criterion = nn.CrossEntropyLoss()  # (applies Softmax)



softmax numpy: [0.65900114 0.24243297 0.09856589]
softmax torch: tensor([0.6590, 0.2424, 0.0986])
Loss1 numpy: 0.3567
Loss2 numpy: 2.3026
PyTorch Loss1: 0.4170
PyTorch Loss2: 1.8406
Actual class: 0, Y_pred1: 0, Y_pred2: 1
Batch Loss1:  0.2834
Batch Loss2: 1.6418
Actual class: tensor([2, 0, 1]), Y_pred1: tensor([2, 0, 1]), Y_pred2: tensor([0, 2, 0])
